proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,634 @@
+       simply filter out bindings that refer to the wrong kind of entities. 
+
+Qualified names become more structured
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Currently, qualified names consist of a simple name, possibly qualified
+by a module name. That is, they could be represented by a type like:
+
+::
+
+   data QualName = Qualified ModuleName Name | Simple Name
+
+We propose to replace this with a type like:
+
+::
+
+   data QualName = Qualified QualName Name | Simple Name
+
+That is, a qualified name is a simple name with a (possibly empty) list
+of qualifying names. These correspond to the dot-separated segments of
+the textual name.
+
+Scopes contain simple names only
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Currently, the ``inScope`` relation relates *qualified* names to
+entities, with the following signature:
+
+::
+
+   inScope :: Rel Module QualName Entity
+
+We propose that instead it relates *simple* names to entities, and we
+add a new relation ``refersTo`` which relates qualified names to
+entities in a scope. [#]_
+
+.. [#] By analogy to programming languages: the bindings in scope are like an environment,
+   simple names are like variables, qualified names are like expressions, and ``refersTo``
+   is ``evalExpr``.
+
+::
+
+   inScope :: Rel Module Name Entity
+   refersTo :: Rel Module QualName Entity
+
+``refersTo`` works as follows:
+
+- If the qualified name has no
+  qualifiers, then it refers to whatever that simple name maps to in the
+  scope.
+- If the qualified name has a qualifier, then we work out what
+  the qualifying name refers to. If it refers to a module, then we look at
+  the bindings exported by the module to find one matching the name.
+
+That is, we interpret ``A.b`` by first working out what ``A`` refers to, as a
+module, and then looking inside it to find ``b``. ``A`` may itself be
+qualified (since modules may export module bindings, just like any other
+bindings), hence the need for the recursive lookup.
+
+Exporting module names
+^^^^^^^^^^^^^^^^^^^^^^
+
+Module names *cannot* be exported, since there are no export specifiers that can
+refer to them. [#]_ The existing ``module M`` export specifier
+continues to work as it is defined in the Haskell report.
+
+.. [#] See `Module export specifiers`_ for the changes to allow this.
+
+Imports
+^^^^^^^
+
+Imports at the top level of a file module are treated specially. Their
+target is interpreted as a *dotted* name, which is resolved using the
+current rules for locating external Haskell modules. [#]_
+
+.. [#] For imports in other settings, see `Local modules`_.
+
+Importing a module brings all the exported bindings of the target module
+into scope, as described in [hsmods]_. It also brings
+the target module into scope bound to its declared simple name (which may be
+a dotted name if it is a file module).
+
+If a module is imported ``qualified`` then the target module is no
+brought into scope.
```

Comment:
@michaelpj, if the intent is to replicate the status quo with the new terminology, then I think it should say:

```
If a module is imported ``qualified`` then the bindings of the target module are not brought into scope directly.
```

Reply:
Whoops, wrote the opposite of what I meant.